internal/guest/network/netns.go

```diff
@@ -216,20 +164,118 @@ func NetNSConfig(ctx context.Context, ifStr string, nsPid int, adapter *prot.Net
 		if err != nil {
 			cos = string(co)
 		}
-		log.G(ctx).Debugf("udhcpc succeeded: %s", cos)
+		entry.Debugf("udhcpc succeeded: %s", cos)
 	}
 
 	// Add some debug logging
-	curNS, _ := netns.Get()
-	// Refresh link attributes/state
-	link, _ = netlink.LinkByIndex(link.Attrs().Index)
-	attr := link.Attrs()
-	addrs, _ := netlink.AddrList(link, 0)
+	if entry.Logger.GetLevel() >= logrus.DebugLevel {
+		curNS, _ := netns.Get()
+		// Refresh link attributes/state
+		link, _ = netlink.LinkByIndex(link.Attrs().Index)
+		attr := link.Attrs()
+		addrs, _ := netlink.AddrList(link, 0)
+		addrsStr := make([]string, 0, len(addrs))
+		for _, addr := range addrs {
+			addrsStr = append(addrsStr, fmt.Sprintf("%v", addr))
+		}
+
+		entry.WithField("addresses", addrsStr).Debugf("%v: %s[idx=%d,type=%s] is %v",
+			curNS, attr.Name, attr.Index, link.Type(), attr.OperState)
+	}
+
+	return nil
+}
 
-	log.G(ctx).Debugf("%v: %s[idx=%d,type=%s] is %v", curNS, attr.Name, attr.Index, link.Type(), attr.OperState)
-	for _, addr := range addrs {
-		log.G(ctx).Debugf("  %v", addr)
+func assignIPToLink(ctx context.Context,
+	ifStr string,
+	nsPid int,
+	link netlink.Link,
+	allocatedIP string,
+	gatewayIP string,
+	prefixLen uint8,
+	enableLowMetric bool,
+	metric int,
+) error {
+	entry := log.G(ctx)
+	entry.WithFields(logrus.Fields{
+		"link":      link.Attrs().Name,
+		"IP":        allocatedIP,
+		"prefixLen": prefixLen,
+		"gateway":   gatewayIP,
+		"metric":    metric,
+	}).Trace("assigning IP address")
+	if allocatedIP == "" {
+		return nil
+	}
+	// Set IP address
+	ip, addr, err := net.ParseCIDR(allocatedIP + "/" + strconv.FormatUint(uint64(prefixLen), 10))
+	if err != nil {
+		return errors.Wrapf(err, "parsing address %s/%d failed", allocatedIP, prefixLen)
+	}
+	// the IP address field in addr is masked, so replace it with the original ip address
+	addr.IP = ip
+	entry.WithFields(logrus.Fields{
+		"allocatedIP": ip,
+		"IP":          addr,
+	}).Debugf("parsed ip address %s/%d", allocatedIP, prefixLen)
+	ipAddr := &netlink.Addr{IPNet: addr, Label: ""}
+	if err := netlink.AddrAdd(link, ipAddr); err != nil {
+		return errors.Wrapf(err, "netlink.AddrAdd(%#v, %#v) failed", link, ipAddr)
+	}
+	if gatewayIP == "" {
+		return nil
+	}
+	// Set gateway
+	gw := net.ParseIP(gatewayIP)
+	if gw == nil {
+		return errors.Wrapf(err, "parsing gateway address %s failed", gatewayIP)
 	}
 
+	if !addr.Contains(gw) {
+		// In the case that a gw is not part of the subnet we are setting gw for,
+		// a new addr containing this gw address need to be added into the link to avoid getting
+		// unreachable error when adding this out-of-subnet gw route
+		entry.Debugf("gw is outside of the subnet: Configure %s in %d with: %s/%d gw=%s\n",
+			ifStr, nsPid, allocatedIP, prefixLen, gatewayIP)
+		ml := len(gw) * 8
+		addr2 := &net.IPNet{
+			IP:   net.ParseIP(gatewayIP),
```

Comment:
nit: looks like you've parsed it already on L229.

Reply:
oops
fixed